presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -437,33 +437,7 @@ public FragmentProperties addSourceDistribution(PlanNodeId source, PartitioningH
             requireNonNull(distribution, "distribution is null");
 
             partitionedSources.add(source);
-
-            if (!partitioningHandle.isPresent()) {
-                partitioningHandle = Optional.of(distribution);
-                return this;
-            }
-
-            PartitioningHandle currentPartitioning = partitioningHandle.get();
-
-            // If already system SINGLE or COORDINATOR_ONLY, leave it as is (this is for single-node execution)
-            if (currentPartitioning.equals(SINGLE_DISTRIBUTION) || currentPartitioning.equals(COORDINATOR_DISTRIBUTION)) {
-                return this;
-            }
-
-            if (currentPartitioning.equals(distribution)) {
-                return this;
-            }
-
-            Optional<PartitioningHandle> commonPartitioning = metadata.getCommonPartitioning(session, currentPartitioning, distribution);
-            if (commonPartitioning.isPresent()) {
-                needPartitioningHandleReassignment = true;
-                partitioningHandle = commonPartitioning;
-                return this;
-            }
-
-            throw new IllegalStateException(String.format(
-                    "Cannot overwrite distribution with %s (currently set to %s)",
-                    distribution, currentPartitioning));
+            return setDistribution(distribution, metadata, session);
```

Comment:
There are differences here, although I don't know how significant the difference is.

Difference 1. I don't know why this is in place. It considers hash/arbitrary/broadcast/fixed-passthrough as compatible. I suppose it's ok to leave as is because it doesn't affect either system source partitioning or connector-provided partitioning.

```
            if (isCompatibleSystemPartitioning(distribution)) {
                return this;
            }
```

Difference 2. This seems very weird for two reasons. First, it only allows system source partitioning, not a connector provided one. Secondly, it seems asymmetric. It allow system source partitioning to combine with any other partitioning, but I don't see code that allows the other way around.

```
            if (currentPartitioning.equals(SOURCE_DISTRIBUTION)) {
                this.partitioningHandle = Optional.of(distribution);
                return this;
            }
```